### PR TITLE
Implement Hlms custom pieces swap for Objects & Terrain

### LINKED ISSUE #444: Heightmaps don't work properly with segmentation camera
## Environment
* OS Version: Ubuntu 18.04
* Source or binary build?
  Source: https://github.com/ignitionrobotics/ign-rendering/commit/eb7d6eaa5c6129fd76acbe222f6ca8655599342a

## Description
* Expected behavior: Heightmaps shouldn't be included in the segmentation output if they don't have a label.
* Actual behavior: Heightmaps can be seen if they aren't labeled.

A temporary fix that disables heightmap from the segmentation camera completely has been made in #443, but we should implement a solution that allows users to apply a label to the heightmap if they desire.

## Steps to reproduce

1. Run a world with a heightmap and semantic segmentation sensor in `ign-gazebo` (the fortress demo world has both)
2. View the segmentation output in the gazebo GUI by viewing the relevant colored map segmentation topic
3. You'll see a part of the heightmap in the segmentation output

## Output
partial_heightmap

### LINKED ISSUE #504: Port custom materials to Hlms pieces
## Desired behavior

Right now specific passes [such as the Segmentation Camera](https://github.com/ignitionrobotics/ign-rendering/issues/444) iterates through all objects, changes their material to a custom low level materials, and renders that.

This works but has a few issues:

 - It cannot handle all details that HlmsPbs/HlmsUnlit does (e.g. HW skeletal animation)
 - It doesn't handle custom Hlms implementations (such as Terra's)
 - It can be slow to change the materials so often (switch it on every object to render, then restore them back)
 - It isn't API portable (it needs one material per API i.e. D3D11/Metal/GL)

Ogre +2.1 supports Hlms pieces, which allows users to add custom code that can be run before/during/after our default code runs.

This solves all outstanding issues:

 - It's fast
 - No need to switch materials (we'd still need to iterate through objects to disable those unwanted, unless we use visibility flags instead, but that's out of scope of this ticket)
 - It's API portable
 - It works with (nearly?) all Hlms implementations
 - All features supported by the underlying implementation that affect vertex position (i.e. HW skeletal animation) are automatically supported.

## Alternatives considered

This is the recommended way.

## Implementation suggestion

I'll be the one implementing this feature

### PR DESCRIPTION
**Pull request by darksylinc**
# 🦟 Bug fix

Fixes #504
Fixes #444

## Summary

The 3rd and final PR to ticket #504 !

After this PR:

 - Terra now works with all Cameras (Thermal, SelectionBuffer, Segmentation, Laser) as documented in ticket #444 (backporting this to Fortress w/out breaking ABI could be a ton of work unfortunately)
 - Fixes various misc bugs of wrong values used in these cameras by regular objects when their properties weren't set (documented in #521, technically it fixes that one, but the fix can be backported so the ticket can will open until these versions are fixed or EoL)
 - Deformation works correctly under these cameras (e.g. skeletal animation and morph target)
   - Only exception is Thermal camera when using `heatSignature` texture. This was not ported to Hlms pieces and we remain swapping a regular low level material
 - Swaps the low level material for all cameras, not just a few of them (as documented in #544; however it doesn't fix the underlying issue that we don't respect the original material's vertex shader deformation). This can be fixed in older versions as well (w/out breaking ABI).
 - Increases test coverage by throwing an error (and crashing) if an object tries to render without a color set while in `IORM_SOLID_COLOR` and `IORM_SOLID_THERMAL_COLOR_TEXTURED` modes. Previously Gazebo would render colour to non-colour render targets and the unit tests wouldn't always catch it.

### Notes:
 - Codeconv reports lower coverage partly because this patch now handles low level materials in all cameras; but there are no tests covering these new paths.
     - No tests covering these paths are tracked by #543 and does not just affect Garden, but also Fortress
     - Fortress does not currently handle this case, which would result in incorrect sensor data. This is a bug in Fortress and is being tracked by #544
     - This PR handles the case, but it just reverts to a standard default shader which is suboptimal. Proper solution would be to use the original vertex shader and various solutions to this problem are proposed in #544
 - Codeconv reports lower coverage partly because there are new codepaths to handle the terrain with each sensor (e.g. Laser, Segmented camera, etc) but no unit test is testing these paths.
     - Ticket #546 was created to track this issue

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: ogre2/src/Ogre2IgnHlmsTerraPrivate.cc [hunk 1]
```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
```
Comment (iche033) [inline ogre2/src/Ogre2IgnHlmsTerraPrivate.cc:2]: 2022?

**Review by iche033 — COMMENTED** [review]
lots of stuff to go through :) I did a quick first pass and some testing. 

I created [sensor_heightmap.sdf](https://gist.github.com/iche033/fdbb3423a316ff37f1b1e24861f5b149) world file for testing in ignition gazebo and confirms that heightmap now appears in segmentation camera, lidar, and thermal camera (using fixed temperature). Looks promising!

sensor_heightmap

File: ogre2/src/Ogre2SegmentationMaterialSwitcher.hh [hunk 2]
```diff
@@ -67,6 +69,14 @@ class IGNITION_RENDERING_OGRE2_VISIBLE Ogre2SegmentationMaterialSwitcher :
   /// \return The map between color and label IDs
   public: const std::unordered_map<int64_t, int64_t> &ColorToLabel() const;
 
+  /// \brief Create a color to apply for the given visual
+  /// \param[in] _visual Visual will be applying the color to
+  /// \param[in,out] _prevParentName A persistent string between call
+  /// to ensure multilink visuals receive the same color
+  /// \return The color to apply to the visual
+  private: Ogre::Vector4 GetColorForVisual(const VisualPtr &_visual,
+                                           std::string &_prevParentName);
```
[the remark this answers is not in the document]
Comment (darksylinc) [reply ogre2/src/Ogre2SegmentationMaterialSwitcher.hh]: Done (all 3 suggestions)

**Review by iche033 — APPROVED** [review]
Did more tests and confirmed that particles are working fine (testing using [sensors_particle_scarttering2.sdf](https://gist.github.com/iche033/0527a92bcae36f0fd85feca36551383a)), and lidar spherical near clipping is also working (tesed using [robots_near_clip.sdf](https://gist.github.com/iche033/7b52dfece48168f90a83b33ebd76aafa)).

**Pull request merged**